tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
         matrix[-1, :] = [0, 0, 1]
 
         return matrix
+
+    @staticmethod
+    def get_tissue_regions(
+        fixed_image: np.ndarray,
+        fixed_mask: np.ndarray,
+        moving_image: np.ndarray,
+        moving_mask: np.ndarray,
+    ) -> Tuple[np.array, np.array, np.array, np.array, tuple]:
+        """Extract tissue region.
+
+        This function uses binary mask for extracting tissue
+        region from the image.
+
+        Args:
+            fixed_image (:class:`numpy.ndarray`):
+                A fixed image.
+            fixed_mask (:class:`numpy.ndarray`):
+                A binary tissue mask for the fixed image.
+            moving_image (:class:`numpy.ndarray`):
+                A moving image.
+            moving_mask (:class:`numpy.ndarray`):
+                A binary tissue mask for the moving image.
+
+        Returns:
+            tuple:
+                - np.ndarray - A cropped image containing tissue region.
+                - np.ndarray - A cropped image containing tissue mask.
+                - np.ndarray - A cropped image containing tissue region.
+                - np.ndarray - A cropped image containing tissue mask.
+                - tuple - Bounding box (min_row, min_col, max_row, max_col).
+
+        """
+        if len(fixed_mask.shape) != 2:
+            fixed_mask = fixed_mask[:, :, 0]
+        if len(moving_mask.shape) != 2:
+            moving_mask = moving_mask[:, :, 0]
+
+        fixed_mask = np.uint8(fixed_mask > 0)
+        moving_mask = np.uint8(moving_mask > 0)
+
+        regions = regionprops(fixed_mask)
+        fixed_minr, fixed_minc, fixed_maxr, fixed_maxc = regions[0].bbox
+        regions = regionprops(moving_mask)
+        moving_minr, moving_minc, moving_maxr, moving_maxc = regions[0].bbox
+        minc, maxc, minr, maxr = (
+            np.min([fixed_minc, moving_minc]),
+            np.max([fixed_maxc, moving_maxc]),
+            np.min([fixed_minr, moving_minr]),
+            np.max([fixed_maxr, moving_maxr]),
+        )
+
+        fixed_tissue_image = fixed_image[minr:maxr, minc:maxc]
+        fixed_tissue_mask = fixed_mask[minr:maxr, minc:maxc]
+        moving_tissue_image = moving_image[minr:maxr, minc:maxc]
+        moving_tissue_mask = moving_mask[minr:maxr, minc:maxc]
+        moving_tissue_image[np.all(moving_tissue_image == (0, 0, 0), axis=-1)] = (
+            243,
+            243,
+            243,
+        )
+        return (
+            fixed_tissue_image,
+            fixed_tissue_mask,
+            moving_tissue_image,
+            moving_tissue_mask,
+            (minr, minc, maxr, maxc),
+        )
+
+    @staticmethod
+    def find_points_inside_boundary(mask: np.ndarray, points: np.ndarray):
```

Comment:
I understand similar functions have been implemented in patch extraction.

Reply:
https://tia-toolbox.readthedocs.io/en/latest/_modules/tiatoolbox/wsicore/wsireader.html#VirtualWSIReader

Reply:
Now using PatchExtractor.filter_coordinates_fast.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L647-L673